File: golang/toeplitz_matrix.go
package main

/*
Space	: O(m + n)
Time	: O(mn)
*/

func isToeplitzMatrix(matrix [][]int) bool {
	d := make(map[int]int)
	m, n := len(matrix), len(matrix[0])

	for y := 0; y < m; y++ {
		for x := 0; x < n; x++ {
			if _, ok := d[x-y]; ok {
				if d[x-y] != matrix[y][x] {
					return false
				}
			} else {
				d[x-y] = matrix[y][x]
			}
		}
	}

	return true
}

File: python3/toeplitz_matrix.py
"""
Space   : O(1)
Time    : O(mn)
"""


class Solution:
    def isToeplitzMatrix(self, matrix: List[List[int]]) -> bool:
        if len(matrix) == 0:
            return True
        if len(matrix[0]) == 0:
            return True

        ly = len(matrix)
        lx = len(matrix[0])

        for y in range(ly-1):
            for x in range(lx-1):
                if matrix[y][x] != matrix[y+1][x+1]:
                    return False

        return True
